pkg/container/watch.go

```diff
@@ -0,0 +1,115 @@
+package container
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/events"
+	"github.com/containerd/typeurl"
+	"github.com/patrickmn/go-cache"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/pkg/stubs"
+)
+
+func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
+	log := log.With().
+		Str("namespace", ns).
+		Str("container", event.ContainerID).Logger()
+
+	log.Debug().Msg("task exited")
+
+	if _, ok := c.failures.Get(event.ContainerID); !ok {
+		c.failures.Set(event.ContainerID, int(0), cache.DefaultExpiration)
+	}
+
+	count, err := c.failures.IncrementInt(event.ContainerID, 1)
+	if err != nil {
+		// this should never happen because we make sure value
+		// is set
+		panic(err)
+	}
+
+	log.Debug().Int("count", count).Msg("recorded stops")
+
+	var reason error
+	if count < failuresBeforeDestroy {
+		log.Debug().Msg("trying to restart the container")
+		<-time.After(restartDelay) // wait for 2 seconds
+		reason = c.start(ns, event.ContainerID)
+	} else {
+		reason = fmt.Errorf("deleting container due to so many crashes")
+	}
+
+	if reason != nil {
+		log.Debug().Msg("deleting container due to so many crashes")
+
+		stub := stubs.NewProvisionStub(c.client)
```

Comment:
We don't create a mew connection. The connection is available on the module structure. It's the stub that is created which is just a wrapper around the connection to expose the module methods. Hence it can be deallocated cleanely after the stack is over.